### PR TITLE
[SPARK-43116][SQL] Fix Cast.forceNullable

### PR DESCRIPTION
**Pull request by jackylee-ch**
### What changes were proposed in this pull request?
1. It should include TimestampNTZType,AnsiIntervalType.
2. ArrayType/MapType/StructType to other data type seems should be true.

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Some updated tests

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala [hunk 1]
```diff
@@ -396,6 +396,22 @@ object Cast extends QueryErrorsBase {
...
+    case (DateType | TimestampType, TimestampNTZType) => false
+    case (_, TimestampNTZType) => true
+    case (TimestampNTZType, DateType | TimestampType) => false
+    case (TimestampNTZType, _) => true
+
+    case (_: IntegralType | _: DecimalType, _: AnsiIntervalType) => false
+    case (_, _: AnsiIntervalType) => true
+    case (_: AnsiIntervalType, _: IntegralType | _: DecimalType) => false
+    case (_: AnsiIntervalType, _) => true
+
```
Comment (wangyum) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala:414]: Could you add a test to test all types? For example:
```scala
    TypeCoercionSuite.allTypes.foreach { dt1 =>
      TypeCoercionSuite.allTypes.foreach { dt2 =>
        Cast.forceNullable(dt1, dt2)
          ...
```

Comment (jackylee-ch) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala:414]: Hm. It is a little hard to add a test for this, as the nullable cases encountered by different type conversions are different. Any better idea for this?

Comment (LuciferYang) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala:414]: > Hm. It is a little hard to add a test for this, as the nullable cases encountered by different type conversions are different. Any better idea for this?

friendly ping @wangyum , any better suggestion?

**Comment by LuciferYang** [comment]
How should we deal with this PR? @wangyum

**Pull request closed without merging**